nearlib/account.js

```diff
@@ -1,39 +1,85 @@
+const ed25519 = require('ed25519');
+const bs58 = require('bs58');
+const MAX_RETRIES = 3;
 
-/**
- * Creates a new account with a random new key.
- */
-exports.createRandomAccount = async function() {
-    // Not implemented yet
-};
+class Account {
+    constructor(keyStore) {
+        this.keyStore = keyStore;
+    }
 
-/**
- * Retrieves account data by plain-text account id. 
- */
-exports.viewAccount = async account_id => {
-    return await viewAccount(account_id);
-};
+    /**
+     * Creates a new account with a given name and key,
+     */
+    async createAccount (newAccountId, publicKey, amount, originatorAccountId) {
+        const createAccountParams = {
+            originator: originatorAccountId,
+            new_account_id: newAccountId,
+            amount: amount,
+            public_key: publicKey,
+        };
 
+        const transactionResponse = await this.submitTransaction("create_account", createAccountParams);
+        return transactionResponse;
+    };
 
-// Implementation
-const superagent = require('superagent');
+    /**
+     * Retrieves account data by plain-text account id. 
+     */
+    async viewAccount (account_id) {
+        return await viewAccount(account_id);
+    };
 
-const viewAccount = async account_id => {
-    const viewAccountResponse = await request('view_account', {
-        account_id: account_id,
-    });
-    return viewAccountResponse;
-}
+    async viewAccount (account_id) {
+        const viewAccountResponse = await this.request('view_account', {
+            account_id: account_id,
+        });
+        return viewAccountResponse;
+    }
 
-const request = async (methodName, params) => {
-    try {
-        // TODO: (issue 320) make this configurable and testable
-        const response = await superagent
-            .post(`http://localhost:3030/${methodName}`)
-            .use(require('superagent-logger'))
-            .send(params);
-        return JSON.parse(response.text);
-    } catch(e) {
-        console.error("error calling rpc ", e);
-        throw e;
+    async signTransaction (transaction, sender) {
+        const stringifiedTxn = JSON.stringify(transaction);
+        const encodedKey = this.keyStore.getKey(sender).secret_key;
+        const key = bs58.decode(encodedKey);
+        const signature = [...ed25519.Sign(new Buffer(stringifiedTxn, 'utf8'), key)];
+        const response = {
+            body: transaction,
+            signature: signature
+        };
+        return response;
+    };
+
+    async submitTransaction (method, args) {
+        // TODO: Make sender param names consistent
+        // TODO: https://github.com/nearprotocol/nearcore/issues/287
+        const senderKeys = ['sender_account_id', 'originator_account_id', 'originator_id', 'sender', 'originator'];
+        const sender = senderKeys.map(key => args[key]).find(it => !!it)
+        const nonce = await this.getNonce(sender);
+        for (let i = 0; i < MAX_RETRIES; i++) {
```

Comment:
Retries were a workaround over buggy JSON-RPC server dropping requests. I think we don't need them anymore.